### PR TITLE
[Cadence 1.0] Optionally check atree storage health before migration

### PR DESCRIPTION
**Pull request by turbolent**
Work towards onflow/cadence#3192

The atree storage health check after the migration might error because the input was already unhealthy.

Optionally check the health before the migration, and do not error if the output is unhealthy when the input was already unhealthy.

### CODE REVIEW

**Review by fxamacker — COMMENTED** [review]
Nice!  This feature can be very useful for troubleshooting.

I left some comments.  The before-migration storage health check needs to happen before `migration.MigrateAccount()`.

File: cmd/util/ledger/migrations/cadence_values_migration.go [hunk 1]
```diff
@@ -124,14 +125,31 @@ func (m *CadenceBaseMigrator) MigrateAccount(
...
+	if m.checkStorageHealthBeforeMigration {
+		storageHealthErrorBefore = storage.CheckHealth()
+		if storageHealthErrorBefore != nil {
+			m.log.Warn().
+				Err(storageHealthErrorBefore).
+				Str("account", address.Hex()).
+				Msg("storage health check before migration failed")
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (turbolent) [reply cmd/util/ledger/migrations/cadence_values_migration.go]: Good points! 

I tried to address them in [d9a1bcf](https://github.com/onflow/flow-go/pull/5591/commits/d9a1bcf98d3d0c07efce32c8ed305512905ca5b6) and [a706f2a](https://github.com/onflow/flow-go/pull/5591/commits/a706f2aef3157bf2a56c593bc28cb12954efda15).

Can you please have another look @fxamacker?

File: cmd/util/ledger/migrations/cadence.go [hunk 2]
```diff
@@ -260,6 +262,7 @@ func NewCadence1ValueMigrations(
 			rwf,
 			diffMigrations,
 			logVerboseDiff,
+			checkStorageHealthBeforeMigration,
```
[the remark this answers is not in the document]
Comment (turbolent) [reply cmd/util/ledger/migrations/cadence.go:265]: Good idea! I'll apply these and add comments

[the remark this answers is not in the document]
Comment (turbolent) [reply cmd/util/ledger/migrations/cadence.go]: Done in 902b1893712d5b106e5e463f6c187733532a5b6e

File: cmd/util/ledger/migrations/cadence.go [hunk 3]
```diff
@@ -246,33 +247,47 @@ func NewCadence1ValueMigrations(
...
+				errorMessageHandler,
+				contracts,
+				capabilityMapping,
+			)
+		},
 	} {
+		accountBasedMigration := migrationConstructor(
+			// Only check storage health before the first migration
+			checkStorageHealthBeforeMigration && index == 0,
+		)
```
Comment (fxamacker) [inline cmd/util/ledger/migrations/cadence.go:289]: Nice ! :+1:

**Review by fxamacker — APPROVED** [review]
LGTM! :+1:

**Review by SupunS — APPROVED** [review]

**Pull request merged**